### PR TITLE
Fix #1385

### LINKED ISSUE #1385: CommandLineStringSplitter bug parsing commandline
I use the code below to parse my commandLine input
```c#
var commandLine = "POST --raw='{""Id"":1,""Name"":""Alice""}'";
CommandLineStringSplitter.Instance.Split(commandLine).ToArray();
```

The result I got is :

image

Expected result:

image

### PR DESCRIPTION
**Pull request by krsahab**
Fix for issue #1385 
Ignore whitespace inside the quote
Ignore quotes if seeking == Boundary.WordEnd
Tests added

### CODE REVIEW

File: src/System.CommandLine/Parsing/CommandLineStringSplitter.cs [hunk 1]
```diff
@@ -25,52 +27,66 @@ public IEnumerable<string> Split(string commandLine)
             var pos = 0;
 
             var seeking = Boundary.TokenStart;
-            int? skipQuoteAtIndex = null;
+            var seekingQuote = Boundary.QuoteStart;
+            StringBuilder sb = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (krsahab) [reply src/System.CommandLine/Parsing/CommandLineStringSplitter.cs]: StringBuilder removed and code modified accordingly

**Review by jonsequitur — APPROVED** [review]

File: src/System.CommandLine/Parsing/CommandLineStringSplitter.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonsequitur) [reply src/System.CommandLine/Parsing/CommandLineStringSplitter.cs]: Beautiful! Nice work!

**Comment by jonsequitur** [comment]
Thanks, @krsahab!

File: src/System.CommandLine.Tests/Parsing/CommandLineStringSplitterTests.cs [hunk 2]
```diff
@@ -75,5 +75,25 @@ public void It_handles_multiple_options_with_quoted_arguments()
...
         }
+
+        [Fact]
+        public void Internal_quotes_do_not_cause_string_to_be_split()
+        {
+            var commandLine = @"POST --raw='{""Id"":1,""Name"":""Alice""}'";
+
+            _splitter.Split(commandLine)
+                     .Should()
+                     .BeEquivalentTo("POST", "--raw='{Id:1,Name:Alice}'");
```
Comment (database64128) [inline src/System.CommandLine.Tests/Parsing/CommandLineStringSplitterTests.cs:86]: This is wrong. The double quotes should be preserved.

**Comment by database64128** [comment]
In scenarios such as a simple REPL (https://github.com/database64128/DotnetPlayground/blob/main/CommandLinePlayground/Program.cs#L35), one would expect that each argument was parsed as is. Currently all double quotes are striped away. An input string such as `--plugin '{"validJsonFields":"RequireDoubleQuotes"}'` would become `'{validJsonFields:RequireDoubleQuotes}'`. Thie single quotes should have been striped away. The double quotes should have been preserved.

File: src/System.CommandLine.Tests/Parsing/CommandLineStringSplitterTests.cs (hunk 2 above)
Comment (krsahab) [reply src/System.CommandLine.Tests/Parsing/CommandLineStringSplitterTests.cs:86]: Minimal Changes have been done just to fix the mentioned issue and not to modify any existing logic.
It was preexisting logic to stripe away all the double-quotes. I tried to preserve it but it was conflicting with other test cases.

**Comment by krsahab** [comment]
As per preexisting logic, only double quotes should be removed, not any other characters (whitespace is a special case). If single quotes should be removed, then the question is What all characters should be removed or preserved?

**Comment by database64128** [comment]
@krsahab Thank you for the explanation! Maybe I should open a new issue for my proposed changes.

> If single quotes should be removed, then the question is What all characters should be removed or preserved?

I think the behavior should be aligned with popular shells: Remove the outermost pair of quotes.

**Comment by jonsequitur** [comment]
The behavior is intended to match this: https://docs.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw. I'm sure we've missed some details.

This is _separate from_ the shell behavior. Different shells have their own escaping rules, so in many cases the application entry point might never even see the quotes you typed into the terminal. `CommandLineToArgvW` then has additional behaviors that will strip out quotes. People generally see the end result after the command line passes through both of these layers, which is part of what makes this so confusing.

**Pull request merged**